python/cudf/cudf/tests/test_joining.py

```diff
@@ -1123,11 +1124,21 @@ def test_typecast_on_join_overflow_unsafe(dtypes):
     lhs = cudf.DataFrame({"a": [1, 2, 3, 4, 5]}, dtype=dtype_l)
     rhs = cudf.DataFrame({"a": [1, 2, 3, 4, dtype_l_max + 1]}, dtype=dtype_r)
 
-    with pytest.warns(
+    p_lhs = lhs.to_pandas()
+    p_rhs = rhs.to_pandas()
+
+    with expect_warning_if(
+        (dtype_l.kind == "f" and dtype_r.kind in {"i", "u"})
+        or (dtype_l.kind in {"i", "u"} and dtype_r.kind == "f"),
         UserWarning,
-        match=(f"Can't safely cast column" f" from {dtype_r} to {dtype_l}"),
     ):
-        merged = lhs.merge(rhs, on="a", how="left")  # noqa: F841
+        expect = p_lhs.merge(p_rhs, on="a", how="left")
+    got = lhs.merge(rhs, on="a", how="left")
+
+    # The dtypes here won't match exactly because pandas does some unsafe
```

Comment:
In pandas, we raise a UserWarning too when a user tries to merge on an int and float column but still allow the operation. Maybe we should suggest the user that they cast their columns to avoid the warning

Reply:
I think the other issue I had with the pandas operation is the output dtype it chooses isn't safe (hence `check_dtypes=False` below). I've forgotten exactly what output type I got, it might have inferred an integer type?

Reply:
Ah yes thats true

```
In [5]: lhs = pd.DataFrame({"a": range(5)}, dtype="int64")

In [6]: rhs = pd.DataFrame({"b": [1, 2, 3, 4, np.iinfo(np.int64).max + 1]}, dtype="float64")

In [9]: lhs.merge(rhs, on="a", how="left").dtypes
<ipython-input-9-267702b4848a>:1: UserWarning: You are merging on int and float columns where the float values are not equal to their int representation.
  lhs.merge(rhs, on="a", how="left").dtypes
Out[9]:
a    int64
dtype: object
```

I guess technically it's safer if the resulting type was float even thought the values are the "same" and `how="left"` has integer type

Reply:
I guess that it's reasonable to assume that if `how="left"` we should use the dtype of the left column. Joining on a floating point-typed column is pretty suspect in any case. @shwina what do you think we should do in this case? Maybe it does make sense to convert to int. @mroeschke does pandas do any introspection to determine whether the values in the column are equivalent to integers (e.g. via `x == x.round()`) or does it just cast and hope for the best?

Reply:
> does pandas do any introspection to determine whether the values in the column are equivalent to integers (e.g. via x == x.round()) or does it just cast and hope for the best?

In this code path, it's a naive cast-left-to-right's-dtype and `==` to see if they are equivalent.

More generally we have a "can this scalar fit in this dtype" function https://github.com/pandas-dev/pandas/blob/dba2080af7647145ce82669c9d67cf2408b2a48a/pandas/core/dtypes/cast.py#L1642. I think we might have an array version of this but can't find it at the moment

Reply:
The "can this scalar fit in this dtype" approach is a non-starter for cudf due to the level of introspection it would require and the associated performance bottlenecks. The naive cast and check is a potentially acceptable though. @brandon-b-miller what do you think?

Reply:
Originally we wanted to [match what pandas did here](https://github.com/rapidsai/cudf/issues/2230#issuecomment-555562743), but there were some edge cases I ran into at the time it seemed like we shouldn't try and match. This led to us wanting to [have our own set of rules](https://github.com/rapidsai/cudf/issues/2230#issuecomment-555578556).

Some of what we came up with _did_ involve data introspection. For instance as you said, for a left join it seemed to make sense to prioritize the dtype of the left table and cast the other operand. IIRC it was while pondering this that I realized the danger that both rounding and overflow errors present in this scheme. For example if the left join column is int dtype and the right is float, we'd match a `1` on the left with both a `1.0` and a `1.5` on the right and end up with too many rows in the result. IIRC, this is what led us to `can_cast_safely`, which is supposed to tell us if every value in that column can be fit into a target dtype losslessly.

Reply:
Got it. It looks like a lot of thought already went into accepting this divergence from pandas, so I'm fine with leaving this as is. Happy to revisit in the future if anyone else feels strongly.